fix: refresh slots on timezone change for booker timezone restrictions

hackice20 opened the pull request:
## What does this PR do?

Refreshes available slots when booker changes timezone on events with restriction schedules that use booker's timezone.
Previously slots weren't refreshing because conversion was client-side only. Now we detect timezone changes and refetch slots when needed.
Only refreshes when all conditions are met:
Timezone changed
Event has `restrictionScheduleId`
Event has `useBookerTimezone: true`
No unnecessary API calls for regular events.

- Fixes #22319 

## Visual Demo (For contributors especially)
#### Video Demo (if applicable):

https://github.com/user-attachments/assets/3eb9ca24-22a7-4578-bcad-27d128ce50c0


## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
1. Create event type with a restriction schedule and enable "Use booker's timezone"
2. Open booking page in incognito
3. Open DevTools → Network tab, filter by slots
4. Change timezone in the booker
5. Expected: New getSchedule API call fires, slots refresh

## Checklist

- I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code does follow the style guidelines of this project
- I have commented my code, particularly in hard-to-understand areas
- I have checked if my changes generate no new warnings

hackice20 committed:
fix: refresh slots on timezone change for booker timezone restrictions

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=26026) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=26026) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=26026) it.</sub>

Ryukemeister commented on packages/features/bookings/Booker/components/hooks/useTimezoneChangeDetection.ts:
```diff
@@ -0,0 +1,52 @@
+import { useEffect, useRef } from "react";
```
hi @hackice20, thanks for raising this PR. I love the tests that you've added for the custom hook, however there are a few suggestions i'd like to give: instead of having some many useEffects why dont we memoize this hook and make it something like this

```
import { useMemo, useRef, useEffect } from "react";
import { useBookerTime } from "./useBookerTime";

type TimezoneChangeDetectionEvent = {
  restrictionScheduleId?: number | null;
  useBookerTimezone?: boolean;
} | null | undefined;

export const useTimezoneChangeDetection = (eventData: TimezoneChangeDetectionEvent) => {
  const { timezone } = useBookerTime();
  
  // this keeps track of the previous timezone across renders
  const prevTimezoneRef = useRef<string>(timezone);
  const prevTimezone = prevTimezoneRef.current;

  // detects changes during the current render
  const hasTimezoneChanged = prevTimezone !== timezone;

  // udate the ref after the render logic is calculated
  useEffect(() => {
    prevTimezoneRef.current = timezone;
  }, [timezone]);

  return useMemo(() => {
    const hasRestrictionSchedule = Boolean(eventData?.restrictionScheduleId);
    const isUsingBookerTimezone = Boolean(eventData?.useBookerTimezone);

    // this matches the original logic: It is only true during the render
    // where the timezone has actually flipped.
    const shouldRefreshSlots = hasTimezoneChanged && hasRestrictionSchedule && isUsingBookerTimezone;

    return {
      shouldRefreshSlots,
      currentTimezone: timezone,
      previousTimezone: prevTimezone,
    };
    // We include hasTimezoneChanged here so the memoized object updates 
    // correctly when the flip occurs.
  }, [timezone, hasTimezoneChanged, eventData?.restrictionScheduleId, eventData?.useBookerTimezone]);
};
```

this way we preserve the original logic while also making sure we're memoizing places where we can, what do you think about this?

hackice20 committed:
refactor: use useMemo for timezone change detection

hackice20 committed:
revert: remove unnecessary formatting changes

hackice20 commented:
Hey @Ryukemeister ,Thanks for the suggestion! Refactored to use `useMemo` and simplified the logic as you suggested. All tests pass and the hook is now more efficient. Lmk what you think

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 commented:
@Ryukemeister

hackice20 commented:
@alishaz-polymath

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

Ryukemeister reviewed (commented):
hi @hackice20, I saw another PR being created that was being referenced as a fix for this PR. i'm a bit confused, do you know what that is about? is it an edge case that we might be missing here or is it something different altogether? can you please work with @MaheshX18 to get this resolved, thank you!

hackice20 commented:
> hi @hackice20, I saw another PR being created that was being referenced as a fix for this PR. i'm a bit confused, do you know what that is about? is it an edge case that we might be missing here or is it something different altogether? can you please work with @MaheshX18 to get this resolved, thank you!

I'll read the code understand what this is about and get this resolved. NO Worries.

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 commented:
@Ryukemeister lmk what you think

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

alishaz-polymath reviewed (changes requested):
This approach doesn't cover for other entry points such as the EventTypeCalendarViewComponent and BookerPlatformWrapper

hackice20 committed:
feat: add timezone refresh for platform components

Add timezone change detection and slot refresh to BookerPlatformWrapper
and EventTypeCalendarViewComponent to handle restriction schedules with
useBookerTimezone enabled.

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 commented:
@alishaz-polymath can you please review this now ? should be fine

alishaz-polymath reviewed (dismissed):
Minor NIT
Rest looking good

hackice20 commented on apps/web/modules/bookings/components/Booker.tsx:
```diff
@@ -141,6 +142,26 @@ const BookerComponent = ({
+      schedule?.invalidate();
+    }
+  }, [shouldRefreshSlots, schedule]);
+
```
Yes perfect , will get it done today itself.

hackice20 replied:
Yes perfect , will get it done today itself.

hackice20 committed:
refactor: extract timezone slot refresh logic into reusable hook

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

alishaz-polymath commented on packages/platform/atoms/calendar-view/EventTypeCalendarViewComponent.tsx:
```diff
@@ -1,4 +1,4 @@
-import { useMemo } from "react";
+import { useMemo, useEffect } from "react";
```
```suggestion
import { useMemo } from "react";
```
Let's remove the unused `useEffect`

alishaz-polymath reviewed (dismissed):
NITs

hackice20 committed:
Update packages/platform/atoms/calendar-view/EventTypeCalendarViewComponent.tsx

Co-authored-by: Syed Ali Shahbaz <52925846+alishaz-polymath@users.noreply.github.com>

hackice20 committed:
Update packages/platform/atoms/calendar-view/EventTypeCalendarViewComponent.tsx

Co-authored-by: Syed Ali Shahbaz <52925846+alishaz-polymath@users.noreply.github.com>

hackice20 commented:
> Testing this locally now. 🙏

yes

alishaz-polymath committed:
fix

alishaz-polymath reviewed (changes requested):
A 🔴 flag when testing this, irrespective of the 'userbookerTimezone' setting being enabled/disabled a getSchedule request is sent when changing the timezone. We can't have these extra calls unless `useBookerTimezone` is enabled.

hackice20 committed:
fix: prevent unnecessary getSchedule calls when useBookerTimezone is disabled

alishaz-polymath reviewed (dismissed):
Let's go 🚀 🚀 🚀 
Excellent work @hackice20 👏

sean-brydon reviewed (dismissed):
LGTM from a consumer perspective - nice job!

alishaz-polymath committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

emrysal reviewed (dismissed):
Looks good to me from a foundational POV 👍 - thank you!

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

alishaz-polymath committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

alishaz-polymath commented:
@hackice20 Let's fix the type errors

hackice20 committed:
fix: add timezone fields to BookerEvent type

alishaz-polymath committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

alishaz-polymath commented:
@hackice20 Type checks still failing

hackice20 committed:
fix: add missing properties to BookerEvent and BookerEventProfile types

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 commented:
@alishaz-polymath you can merge main branch into this and check, should be fine.
@sean-brydon @emrysal @Ryukemeister my recent commit, dismissed reviews.

alishaz-polymath committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

alishaz-polymath committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

ThyMinimalDev committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 committed:
trying to fix type errors

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change

hackice20 commented:
Hello @alishaz-polymath sorry to bother you again and again, was looking if 
```ts
 // Dynamic group events don't have restriction schedules configured
      restrictionScheduleId: null,
      useBookerTimezone: false,
```
in the `getPublicEvent.ts` file will this work ?

hackice20 committed:
Add restrictionScheduleId and useBookerTimezone fields

hackice20 committed:
Merge branch 'main' into fix/slots-refresh-on-timezone-change
